src/category_theory/limits/over.lean

```diff
@@ -84,6 +86,131 @@ instance forget_preserves_colimits [has_colimits.{v} C] {X : C} :
+      is_limit :=
+        { lift := λ s, over.hom_mk _,
+          fac' := λ _ j, j.elim,
+          uniq' := λ s m _,
+            begin
+              ext,
+              rw over.hom_mk_left,
+              have := m.w,
+              dsimp at this,
+              rwa [category.comp_id, category.comp_id] at this
+            end } } } }
+
+-- TODO: this should work for any connected limit, not just pullbacks
+/-- Given pullbacks in C, we have pullbacks in C/B -/
+instance {B : C} [has_pullbacks.{v} C] : has_pullbacks.{v} (over B) :=
+begin
+  refine ⟨⟨λ F, _⟩⟩,
+  let X : over B := F.obj walking_cospan.one,
+  let Y : over B := F.obj walking_cospan.left,
+  let Z : over B := F.obj walking_cospan.right,
+  let f : Y ⟶ X := (F.map walking_cospan.hom.inl),
+  let g : Z ⟶ X := (F.map walking_cospan.hom.inr),
+  let L : over B := over.mk (pullback.fst ≫ Y.hom : pullback f.left g.left ⟶ B),
+  let π₁ : L ⟶ Y := over.hom_mk pullback.fst,
+  let π₂ : L ⟶ Z, refine @over.hom_mk _ _ _ L Z (pullback.snd : L.left ⟶ Z.left) _,
+    simp,
+    rw [← over.w f, ← category.assoc, pullback.condition, category.assoc,  over.w g],
+  refine {cone := cone.of_pullback_cone (pullback_cone.mk π₁ π₂ _), is_limit := {lift := _, fac' := _, uniq' := _}},
+    ext, simp, erw pullback.condition,
+  intro s,
+  apply over.hom_mk _ _,
+  apply pullback.lift (s.π.app walking_cospan.left).left (s.π.app walking_cospan.right).left,
+  rw ← over.comp_left, rw ← over.comp_left,
+  rw s.w, rw s.w, simp,
+  show pullback.lift (((s.π).app walking_cospan.left).left) (((s.π).app walking_cospan.right).left) _ ≫
+    (pullback.fst ≫ Y.hom : pullback f.left g.left ⟶ B) = (s.X).hom, simp, refl,
+  intros s j, simp, ext1, dsimp,
+  cases j, simp, simp, simp,
+  show _ ≫ (((pullback_cone.mk π₁ π₂ _).π).app walking_cospan.one).left = ((s.π).app walking_cospan.one).left,
+  dunfold pullback_cone.mk, dsimp,
+  show pullback.lift (((s.π).app walking_cospan.left).left) (((s.π).app walking_cospan.right).left) _ ≫
+    pullback.fst ≫ f.left =
+  ((s.π).app walking_cospan.one).left, simp, rw ← over.comp_left, rw ← s.w walking_cospan.hom.inl,
+  intros s m J, apply over.over_morphism.ext, simp, apply pullback.hom_ext,
+  simp at J, dsimp at J,
+  have := J walking_cospan.left, dsimp at this, simp, rw ← this, simp,
+  have := J walking_cospan.right, dsimp at this, simp, rw ← this, simp
+end
+
+/-- Given pullbacks in C, we have binary products in any over category -/
+instance over_has_prods_of_pullback [has_pullbacks.{v} C] (B : C) :
+  has_binary_products.{v} (over B) :=
+{has_limits_of_shape := {has_limit := λ F, over_product_of_pullbacks B F}}
+
+-- A collection of simp lemmas to decompose products in the over category
+@[simp]
+lemma over_prod_is_pullback [has_pullbacks.{v} C] {B : C} (F : discrete walking_pair ⥤ over B) :
+  limits.limit F = @over.mk _ _ B (pullback (F.obj walking_pair.left).hom (F.obj walking_pair.right).hom) (pullback.fst ≫ (F.obj walking_pair.left).hom) := rfl
```

Comment:
I don't think this is a good `simp` lemma. Also below. When are you going to want to use these lemmas automatically?

Reply:
I found the simp lemma `over_prod_map` very useful in https://github.com/b-mehta/topos/blob/master/src/locally_cartesian_closed.lean, and the others useful in proving that. Perhaps not all of them are necessary though.

Reply:
It turned out to suffice to have `over_prod_map` and `over_prod_pair` (which I guess isn't the best name). I'm happy to remove the others if you prefer.

Reply:
I'm not really happy with even those two as `simp` lemmas. It seems their whole point is to spill out the implementation details, making for fragile proofs.

Perhaps you can make them local simp lemmas at the point of use? Or try to adapt those proofs to use more of the API, and less of the implementation details?

Reply:
Alright, I've changed the proofs. The statement of `over_prod_map` is horrible now - do you have suggestions for tidying it?

Reply:
Ah, I see. Okay, rather than any proofs that say `X = over.mk ...`, you should prove separate lemmas for each of the data fields of `over`. So rather have `X.left = ...` and so on. These will be much better simp lemmas, and have cleaner statements because you won't be writing proofs in the statements.

Reply:
I'm happy with this now, so I'll mark as resolved